File: examples/late_chunking.py
import heapq
from embed_anything import EmbedData, EmbeddingModel, TextEmbedConfig, WhichModel

from embed_anything import Dtype, ONNXModel
import numpy as np


model:EmbeddingModel = EmbeddingModel.from_pretrained_hf(
    WhichModel.Jina, model_id="jinaai/jina-embeddings-v2-small-en"
)
config = TextEmbedConfig(
    chunk_size=1000,
    batch_size=8,
    splitting_strategy="sentence",
    late_chunking=False,
)

# Embed a single file
data: list[EmbedData] = model.embed_file("test_files/attention.pdf", config=config)

# Print the embedded data
# for d in data:
#     print(d.text)
#     print("---" * 20)

query = "What are positional encodings?"

query_embedding = np.array(model.embed_query([query])[0].embedding)

embedding_array = np.array([e.embedding for e in data])

similarities = np.matmul(query_embedding, embedding_array.T)

# get top 5 similarities and its index
top_5_similarities = np.argsort(similarities)[-10:][::-1]

# Print the top 5 similarities with sentences
for i in top_5_similarities:
    print(f"Score: {similarities[i]:.2} | {data[i].text}")
    print("---" * 20)

File: rust/examples/late_chunking.rs
use std::sync::Arc;

use embed_anything::{config::TextEmbedConfig, embeddings::embed::EmbedderBuilder};

#[tokio::main]
async fn main() {
    let model = Arc::new(
        EmbedderBuilder::new()
            .model_id(Some("jinaai/jina-embeddings-v2-small-en"))
            .revision(None)
            .path_in_repo(Some("model.onnx"))
            .from_pretrained_onnx()
            .unwrap(),
    );

    let config = TextEmbedConfig::default()
        .with_chunk_size(1000, Some(0.3))
        .with_batch_size(4)
        .with_buffer_size(32)
        .with_late_chunking(true);

    let out = model
        .embed_file("test_files/test.pdf", Some(&config), None)
        .await
        .unwrap()
        .unwrap();

    for d in out {
        println!("{}", d.text.unwrap());
        println!("---");
    }
}
